File: leetcode/105.construct-binary-tree-from-Inorder-and-preorder-traversal/go/buildtree.go
package constructbinarytreefrominorderandpreordertraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 || len(preorder) != len(inorder) {
		return nil
	}

	return traversal(preorder, 0, len(preorder), inorder, 0, len(inorder))
}

func traversal(preorder []int, preLeft int, preRight int, inorder []int, inLeft int, inRight int) *TreeNode {
	if preLeft >= preRight {
		return nil
	}

	rootVal := preorder[preLeft]
	root := &TreeNode{Val: rootVal}

	splitIdx := findIdx(inorder, inLeft, inRight, rootVal)
	leftLen := splitIdx - inLeft

	root.Left = traversal(preorder, preLeft+1, preLeft+1+leftLen, inorder, inLeft, splitIdx)
	root.Right = traversal(preorder, preLeft+1+leftLen, preRight, inorder, splitIdx+1, inRight)

	return root
}

func findIdx(nums []int, left int, right int, target int) int {
	for i := left; i < right; i++ {
		if nums[i] == target {
			return i
		}
	}
	return -1
}

File: leetcode/105.construct-binary-tree-from-Inorder-and-preorder-traversal/cpp/buildtree.cpp
#include <vector>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *buildTree(std::vector<int> &preorder, std::vector<int> &inorder)
    {
        if (inorder.empty() || inorder.size() != preorder.size())
            return nullptr;

        return traversal(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }

private:
    TreeNode *traversal(std::vector<int> &preorder, int preLeft, int preRight, std::vector<int> &inorder, int inLeft, int inRight)
    {
        if (preLeft >= preRight)
            return nullptr;

        auto rootVal = preorder[preLeft];
        auto root = new TreeNode(rootVal);

        auto splitIdx = findIdx(inorder, inLeft, inRight, rootVal);
        auto leftLen = splitIdx - inLeft;

        root->left = traversal(preorder, preLeft + 1, preLeft + 1 + leftLen, inorder, inLeft, splitIdx);
        root->right = traversal(preorder, preLeft + 1 + leftLen, preRight, inorder, splitIdx + 1, inRight);

        return root;
    }

    int findIdx(std::vector<int> nums, int left, int right, int target)
    {
        auto it = std::find(nums.begin() + left, nums.begin() + right, target);
        return it == nums.begin() + right ? -1 : it - nums.begin();
    }
};
